File: SSPP/src/sspp/Spiffy.cpp
#include <sspp/Spiffy.h>
#include <sspp/Topic.h>
using namespace std;

namespace SSPP {

string server = SSPP_DEFAULT_TOPIC_SERVER;
string context = ".";

Mutex globalMutex;
map<string,vector<AnySpiffy*> > pushVariables;
map<string,int> writeOnly;

class SpiffyPushService:public Service
{
public:
  SpiffyPushService(const char* host)
  {
    if(!OpenClient(host)) return;
    for(map<string,vector<AnySpiffy*> >::iterator i=pushVariables.begin();i!=pushVariables.end();i++) {
      AnyCollection msg;
      msg["type"] = string("subscribe");
      msg["path"] = i->first;
      SendMessage(msg);
    }
  }
  virtual ~SpiffyPushService() {
    for(map<string,vector<AnySpiffy*> >::iterator i=pushVariables.begin();i!=pushVariables.end();i++) {
      AnyCollection msg;
      msg["type"] = string("unsubscribe");
      msg["path"] = i->first;
      SendMessage(msg);
    }      
  }
  virtual const char* Name() const { return "SpiffyPushService"; }
  virtual const char* Description() const { return ""; }
  virtual bool OnMessage(AnyCollection& message) { 
    ScopedLock lock(globalMutex);
    //push message
    if(pushVariables.size()==1) {
      for(size_t i=0;i<pushVariables.begin()->second.size();i++)
	pushVariables.begin()->second[i]->_updated(message);
    }
    else {
      string path;
      message["path"].as(path);
      map<string,vector<AnySpiffy*> >::iterator it = pushVariables.find(path);
      if(it == pushVariables.end()) {
	cout<<"Invalid subscription message path \""<<path<<"\""<<endl;
	return false;
      }
      for(size_t i=0;i<it->second.size();i++)
	it->second[i]->_updated(message["data"]);
    }
    return true;
  }
  void AddPush(const string& path,AnySpiffy* var) {
    if(pushVariables.count(path) == 0) {
      pushVariables[path] = vector<AnySpiffy*>(1,var);
      //subscribe
      AnyCollection msg;
      msg["type"] = string("subscribe");
      msg["path"] = path;
      SendMessage(msg);
    }
    else
      pushVariables[path].push_back(var);
  }
  void RemovePush(const string& path,AnySpiffy* var)
  {
    Assert(pushVariables.count(path) != 0);
    vector<AnySpiffy*>& vars = pushVariables[path];
    for(size_t i=0;i<vars.size();i++) {
      if(vars[i] == var) {
	vars.erase(vars.begin()+i);
	i--;
      }
    }
    if(vars.empty()) {
      //unsubscribe
      AnyCollection msg;
      msg["type"] = string("unsubscribe");
      msg["path"] = path;
      SendMessage(msg);
      pushVariables.erase(pushVariables.find(path));
    }
  }
};

class SpiffySynchronizedService : public Service
{
public:
  File client;
  int myID;
  SpiffySynchronizedService(const char* host)
  {
    if(!client.Open(host,FILECLIENT)) return;
    //note: possible conflict if multiple services start up simultaneously
    AnyCollection result;
    if(!Get("._spiffy.clientCounter",result)) {
      myID = 0;
    }
    else {
      int nc;
      if(!result.as(nc)) {
	FatalError("Address %s._spiffy.clientCounter not an int\n",host);
      }
      myID = nc;
    }
    Set("._spiffy.clientCounter",AnyCollection(myID+1));
    for(map<string,int>::iterator i=writeOnly.begin();i!=writeOnly.end();i++)
      MarkExclusiveWrite(i->first,true,i->second);
  }
  virtual ~SpiffySynchronizedService() {
    for(map<string,int>::iterator i=writeOnly.begin();i!=writeOnly.end();i++)
      MarkExclusiveWrite(i->first,false,i->second);
  }
  bool MarkExclusiveWrite(const string& path,bool enabled,int id)
  {
    if(enabled) {
      if(!IsWriteEnabled(path,id)) {
	printf("MarkExclusiveWrite(%s,%d,%d): Write permissions for this ID not set\n",path.c_str(),(int)enabled,id);
	return false;
      }
      AnyCollection msg;
      msg["client"] = myID;
      msg["instance"] = id;
      Set(string("._spiffy.writeOnly")+path,msg);
      writeOnly[path] = id;
      return true;
    }
    else {
      if(writeOnly.count(path) == 0) return true;
      if(writeOnly[path] != id) {
	printf("MarkExclusiveWrite(%s,%d,%d): Write permissions for this ID not set\n",path.c_str(),(int)enabled,id);
	return false;
      }
      AnyCollection msg;
      msg["type"] = string("delete");
      msg["path"] = string(".spiffy.writeOnly")+path;
      SendMessage(msg);
      writeOnly.erase(writeOnly.find(path));
      return true;
    }
  }
  bool IsWriteEnabled(const string& path,int id) {
    if(writeOnly.count(path) != 0) {
      if(writeOnly[path] != id) return false;
      else return true;
    }
    AnyCollection msg;
    if(!Get(string("._spiffy.writeOnly")+path,msg)) return true;
    int client,instance;
    if(!msg["client"].as(client) || msg["instance"].as(instance)) return true;
    if(client != myID || id != instance) return false;
    return true;
  }
  bool Set(const string& path,const AnyCollection& value) {
    AnyCollection msg;
    msg["type"] = string("set");
    msg["path"] = path;
    msg["data"] = value;
    SendMessage(msg);    
    return true;
  }
  bool Get(const string& path,AnyCollection& result) {
    AnyCollection msg;
    msg["type"] = string("get");
    msg["path"] = path;
    SendMessage(msg);    
    return WaitForMessage(result);
  }
};

void SetSpiffyServer(const char* addr)
{
  server = addr;
}

const char* GetSpiffyServer()
{
  return server.c_str();
}

bool spiffyStarted = false;
SmartPointer<SpiffySynchronizedService> syncService;
SmartPointer<SpiffyPushService> pushService;
int globalID = 0;

Thread pushThread;
bool pushThreadKill = false;

//destructor
class SpiffyGlobalContext
{
public:
  SpiffyGlobalContext() {}
  ~SpiffyGlobalContext() { if(spiffyStarted) StopSpiffy(); }
};

SpiffyGlobalContext gContext;

bool pushKilled()
{
  return pushThreadKill;
}

void* pushThreadFunc(void*)
{
  pushThreadKill = false;
  RunUntil(*pushService,pushKilled);
  return NULL;
}


bool StartSpiffy(const char* context)
{
  SetSpiffyContext(context);
  pushService = new SpiffyPushService(server.c_str());
  syncService = new SpiffySynchronizedService(server.c_str());
  if(!pushService->Connected() || !syncService->Connected()) {
    printf("Spiffy server on address %s not started\n",server.c_str());
    pushService = NULL;
    syncService = NULL;
    return false;
  }
  spiffyStarted = true;
  pushThread = ThreadStart(pushThreadFunc,NULL);
  return true;
}

void StopSpiffy()
{
  pushThreadKill = true;
  ThreadJoin(pushThread);
  pushService->Disconnect();
  pushService = NULL;
  syncService = NULL;
  spiffyStarted = false;
}

void SetSpiffyContext(const char* ctx)
{
  ScopedLock lock(globalMutex);
  if(ctx == NULL) context="";
  else context=ctx;
}




AnySpiffy::AnySpiffy(const char* _name)
  :name(_name),path(string(context)+"."+string(name)),pushMode(false),newValue(false)
{
  ScopedLock lock(globalMutex);
  id = globalID++;
}

AnySpiffy::~AnySpiffy()
{
  if(writeOnly.count(path) && writeOnly[path]==id) {
    syncService->MarkExclusiveWrite(path,false,id);
  }
  SetPullMode();
}

bool AnySpiffy::SetExclusiveWrite(bool enabled)
{
  ScopedLock lock(globalMutex);
  return syncService->MarkExclusiveWrite(path,enabled,id);
}

bool AnySpiffy::CanWrite()
{
  ScopedLock lock(globalMutex);
  return syncService->IsWriteEnabled(path,id);
}

bool AnySpiffy::CanRead()
{
  return true;
}

void AnySpiffy::SetPullMode()
{
  if(!pushMode) return;
  pushMode = false;
  pushService->RemovePush(path,this);
}
 
void AnySpiffy::SetPushMode()
{
  if(pushMode) return;
  pushMode = true;
  pushService->AddPush(path,this);
  newValue = false;
}

bool AnySpiffy::HasNewValue()
{
  return newValue;
}




bool AnySpiffy::set(const AnyCollection& _value)
{
  ScopedLock lock(globalMutex);
  if(!syncService->IsWriteEnabled(path,id)) return false;
  value = _value;
  return syncService->Set(path,_value);
}

bool AnySpiffy::get(AnyCollection& _value)
{
  if(pushMode) {
    _value = value;
    newValue = false;
    return true;
  }
  else {
    return syncService->Get(path,_value);
  }
}

void AnySpiffy::_updated(const AnyCollection& _value)
{
  value = _value;
  newValue = true;
}


} //namespace SSPP

File: SSPP/Python/sspp/spiffy.py
"""Automatic remote hosting of variables. See spiffy_test.py for an example
of how to use it.

Spiffy servers may perform logging, relay messages between clients, and
manage exclusive writes.

Each process should set a context on startup which will be used as
the namespace for all created variables.  A process can change to a different
context using the spiffy.setContext function.

Usage:
    from sspp import spiffy
    
    #connect to the spiffy server
    spiffy.start((server_ip,server,port),context=".my_context_name")
    
    foo = Spiffy('foo') #creates a variable named foo on the spiffy server under .my_context_name
    print foo.get() #hasn't been set yet, so this returns None
    foo.set(100)
    print foo.get() #will return 100

    #disconnect from the server
    spiffy.stop()
"""

import service
from threading import Thread,Lock
from collections import defaultdict
import time

_map = {}
_lock = Lock()
_context = ""
_pushVariables = defaultdict(list)
_processID = None
_variableCounter = 0
_writeOnly = {}
_pushThread = None
_pushService = None
_syncService = None

def setContext(context=""):
    global _context
    _context = context

def getContext():
    global _context
    return _context

class _PushService(service.Service):
    def __init__(self,host):
        global _map
        service.Service.__init__(self,map=_map)
        self.open(host)
        self.sendMessage({'type':'name','data':'Spiffy push thread'})
        self.sendMessage({'type':'subscribe','path':'._spiffy.null'})
    def checkWriteOnly(self,path):
        self.sendMessage({'type':'subscribe','path':'._spiffy.writeOnly'+path})
    def addPush(self,path,var):
        global _lock,_pushVariables
        with _lock:
            _pushVariables[path].append(var)
            self.sendMessage({'type':'subscribe','path':path})
    def removePush(self,path,var):
        global _lock,_pushVariables
        with _lock:
            _pushVariables[path].erase(var)
            if len(_pushVariables[path])==0:
                self.sendMessage({'type':'unsubscribe','path':path})
    def onMessage(self,msg):
        global _lock,_writeOnly,_pushVariables
        with _lock:
            if msg == None:
                #first message
                return
            if hasattr(msg,'__iter__') and 'path' in msg:
                if msg["path"].startswith('._spiffy.writeOnly'):
                    ofs = len('._spiffy.writeOnly')
                    _writeOnly[msg["path"][ofs:]] = msg["data"]
                else:
                    if msg["path"] not in _pushVariables:
                        print msg["path"]
                        raise RuntimeError("Got a message for a non-push variable")
                    for i in _pushVariables[msg["path"]]:
                        i._update(msg["data"])
            else:
                raise RuntimeError("PushService: invalid push message "+str(msg))

class _SyncService(service.Service):
    def __init__(self,host):
        global _map
        service.Service.__init__(self,map=_map)
        self.open(host)
        self.sendMessage({'type':'name','data':'Spiffy pull thread'})
        self.sendMessage({'type':'get','path':'_spiffy.clientCounter'})
        #print "Waiting for numclients get..."
        result = self.waitForMessage(sameThread=False)
        #print "  done"
        if result==None:
            self.id = 0
        else:
            self.id = result
        self.sendMessage({'type':'set','path':'_spiffy.clientCounter','data':(self.id+1)})

    def isWriteable(self,path,id):
        global _lock,_writeOnly
        _lock.acquire()
        #wait for first writeOnly message to be received
        while True:
            if path in _writeOnly:
                rights = _writeOnly[path]
                _lock.release()
                if rights == None: return True
                return rights['client']==self.id and rights['instance']==id
            _lock.release()
            print "Waiting for access rights for",path
            time.sleep(0.1)
            _lock.acquire()
        _lock.release()
        return True
        
    def markExclusiveWrite(self,path,enabled,id):
        global _lock,_writeOnly
        if enabled:
            if not self.isWriteable(path,id):
                print "Access violation: can't get exclusive write to",path,"for instance ID",id
                return False
            rights = {'client':self.id,'instance':id}
            self.sendMessage({'type':'set','path':'_spiffy.writeOnly'+path,'data':rights})
            with _lock:
                _writeOnly[path] = rights
        else:
            with _lock:
                if path not in _writeOnly:
                    return True
                if _writeOnly[path]['client'] != self.id or _writeOnly[path]['instance'] != id:
                    return True
                del _writeOnly[path]
            self.sendMessage({'type':'delete','path':'_spiffy.writeOnly'+path})
        return True


def _run_push_service(host):
    global _pushService
    _pushService.run()

def _interrupt_handler(signum,frame):
    print "Keyboard interrupt called, joining Spiffy thread..."
    stop()

def start(host=("localhost",4567),context=""):
    global _context,_pushThread,_pushService,_syncService
    _context = context
    _pushService = _PushService(host)
    #the push thread starts the asyncore loop
    _pushThread = Thread(target=_run_push_service,args=(host,))
    _pushThread.start()
    _syncService = _SyncService(host)

def stop():
    global _pushThread,_pushService,_syncService,_writeOnly,_pushVariables
    #unmark any write-only messages, unsubscribe from all messages
    for path,rights in _writeOnly.iteritems():
        if rights['client'] == _syncService.id:
            _pushService.sendMessage({'type':'delete','path':'._spiffy.writeOnly'+path})
    #hopefully the delte messages all get sent within this time?
    time.sleep(0.1)
    _writeOnly = {}
    _pushVariables = {}
    _pushService.terminate()
    _pushThread.join()
    _syncService = None
    _pushService = None

class Spiffy:
    """ A variable on a spiffy server.  Initialize it with a given name and context.

    They can be set to push mode, when the server sends updates, or pull mode, where
    the request is obtained only on a get() call.

    They can be set to exclusive write from a particular Spiffy variable instance
    to exclude other processes/instances from writing to them."""
    def __init__(self,name):
        global _lock,_context,_variableCounter,_pushService
        with _lock:
            self.id = _variableCounter
            _variableCounter += 1
        self.name = name
        self.value = None
        self.path = _context+"."+name
        self.pushMode = False
        _pushService.checkWriteOnly(self.path)
    def set(self,value):
        global _syncService,_pushService
        if not _syncService.isWriteable(self.path,self.id):
            print "Tried writing on an unwritable variable",self.path
            return False
        _syncService.sendMessage({'type':'set','path':self.path,'data':value})
        self.value = value
        return True
    def get(self):
        if self.pushMode:
            return self.value
        global _syncService
        _syncService.sendMessage({'type':'get','path':self.path})
        #print "Waiting for get",self.path,"..."
        self.value = _syncService.waitForMessage(sameThread=False)
        #print "Got",self.path
        return self.value
    def _update(self,value):
        #called by internal push thread for push variables
        self.value = value
    def setPullMode(self):
        global _pushService
        if self.pushMode:
            _pushService.removePush(self.path,self)
            self.pushMode = False
    def setPushMode(self):
        global _pushService
        if not self.pushMode:
            _pushService.addPush(self.path,self)
            self.pushMode = True
    def setExclusiveWrite(self,enabled=True):
        global _syncService
        return _syncService.markExclusiveWrite(self.path,enabled,self.id)
